src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "utils/concepts-enabled.h"
+#include "utils/gate_guard.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/internal/pollable_fd.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/smp.hh>
+
+#include <boost/lockfree/spsc_queue.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+
+namespace v8_engine {
+
+namespace internal {
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() noexcept = 0;
+    virtual void cancel() noexcept = 0;
+    virtual void on_timeout() noexcept = 0;
+    virtual std::chrono::milliseconds get_timeout() const noexcept = 0;
+
+    virtual void done() = 0;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    task(Func&& f, std::chrono::milliseconds timeout)
+      : work_item()
+      , _func(std::forward<Func>(f))
+      , _timeout(timeout) {}
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() noexcept override {
+        try {
+            _func();
+        } catch (...) {
+            exception = std::current_exception();
+        }
+    }
+
+    // Run this function for cancel task execution
+    void cancel() noexcept override { _func.cancel(); }
```

Comment:
cancel is marked as noexcept and there is no try-catch loop inside the excecutor::loop, but _func.cancel() might throw
at least it's not checked by the WrapperFuncForExecutor 

maybe add try-catch here?

Reply:
done